### PR TITLE
extensions: introduce flutter-master

### PR DESCRIPTION
**Pull request by sergiusens**
An extension that sets up an environment with flutter-master and is able
wraps the usage of GNOME into a dependency.

Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/internal/project_loader/_extensions/flutter_master.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                },
+                platform_snap: {
+                    "interface": "content",
+                    "target": "$SNAP/gnome-platform",
+                    "default-provider": "{snap}".format(snap=platform_snap),
+                },
+            },
+            "environment": {"SNAP_DESKTOP_RUNTIME": "$SNAP/gnome-platform"},
+            "layout": {
+                "/usr/bin/gjs": {"symlink": "$SNAP/gnome-platform/usr/bin/gjs"},
```
Comment (kenvandine) [inline snapcraft/internal/project_loader/_extensions/flutter_master.py:94]: We actually don't need the gjs and webkit layouts for flutter

Comment (sergiusens) [reply snapcraft/internal/project_loader/_extensions/flutter_master.py:94]: oooh, nice

**Comment by kenvandine** [comment]
I've come around to the idea of having extensions that match the flutter channels.  We will need both master and dev now and beta and stable later

File: snapcraft/plugins/v1/flutter.py [hunk 2]
```diff
@@ -22,16 +22,15 @@
...
-      Which Flutter channel to use for the build
     - flutter-revision
       (string)
       Which Flutter revision to use for the build. This must be a valid
       revision from the flutter repository.
     - flutter-target
       (string, default: lib/main.dart)
       The main entry-point file of the application
+
+This plugin works best with the flutter related extensions.
```
Comment (cjp256) [inline snapcraft/plugins/v1/flutter.py:33]:
```suggestion
This plugin is designed to work with the flutter extensions.
```

File: tests/unit/project_loader/extensions/test_flutter_master.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            flutter_extension.parts,
+            Equals(
+                {
+                    "gnome-3-28-extension": {
+                        "source": "$SNAPCRAFT_EXTENSIONS_DIR/desktop",
+                        "source-subdir": "gnome",
+                        "plugin": "make",
+                        "build-packages": ["gcc", "libgtk-3-dev"],
+                    },
+                    "flutter-extension": {
```
Comment (cjp256) [inline tests/unit/project_loader/extensions/test_flutter_master.py:94]: Include full name of extension? s/"flutter-extension"/"flutter-master-extension"

Comment (sergiusens) [reply tests/unit/project_loader/extensions/test_flutter_master.py:94]: I would prefer to keep it like this here, just like for gnome and for good reason, we cannot use them together

File: snapcraft/plugins/v1/flutter.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/plugins/v1/flutter.py:33]: this is not necessarily true

File: snapcraft/internal/project_loader/_extensions/flutter_master.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    It tracks the master channel for flutter.
+
+    At build time it ensures the right build dependencies are setup and for
+    the runtime it ensures the application is run in an environment catered
+    for Flutter applications.
+
+    It configures each application with the following plugs:
+
+    \b
```
Comment (hellsworth) [inline snapcraft/internal/project_loader/_extensions/flutter_master.py:38]: why are there these "\b" lines in the comment?

Comment (sergiusens) [reply snapcraft/internal/project_loader/_extensions/flutter_master.py:38]: For help to separate the paragraphs when running "snapcraft help"

**Pull request merged**